### PR TITLE
Improper list patterns in macros

### PR DESCRIPTION
**Pull request by jrvidal**
(This is not quite ready to review yet, I want to do some cleaning and rebasing/squashing first. But it does seem to work and has quite a few tests to back it up)

### CODE REVIEW

File: cogs/syntax-tests.scm [hunk 1]
```diff
@@ -183,6 +183,88 @@
...
+(define-syntax non-list-as-list-multiple
+  (syntax-rules () [
+                  (_ ((a) ... . b) ... ) (quote (b ...))]))
+
+(skip-compile (check-equal? "improper list pattern, nested, collapses to non-list" (non-list-as-list-multiple  "hello" "world") '("hello" "world")))
+
+(skip-compile (define-syntax many-literals
+  (syntax-rules () [
+                  (_ #t ...) 1])))
+
```
Comment (jrvidal) [inline cogs/syntax-tests.scm:264]: A couple of TODOs, but I think this PR is big enough.

File: crates/steel-core/src/parser/expander.rs [hunk 2]
```diff
@@ -391,6 +374,7 @@ impl MacroCase {
 
 #[derive(Clone, PartialEq, Serialize, Deserialize)]
 pub enum MacroPattern {
+    Rest(Box<MacroPattern>),
```
Comment (jrvidal) [inline crates/steel-core/src/parser/expander.rs:377]: :information_source: A new pattern type for anything in the `cdr` of a cons list.

File: crates/steel-core/src/parser/expander.rs [hunk 3]
```diff
@@ -660,72 +652,72 @@ impl MacroPattern {
...
 
-pub fn match_vec_pattern(
-    mut patterns: &[MacroPattern],
-    mut list: &[ExprKind],
-    top_level: bool,
-) -> bool {
-    if top_level {
-        list = &list[1..];
-        patterns = &patterns[1..];
-    }
```
Comment (jrvidal) [inline crates/steel-core/src/parser/expander.rs:693]: :warning: I've reverted this `top_level` gimmick.

File: crates/steel-core/src/parser/expander.rs [hunk 4]
```diff
@@ -735,201 +727,226 @@ pub fn match_vec_pattern(
...
+
+        if !tail_matches {
+            return false;
+        }
+    }
+
+    true
+}
+
+fn match_rest_pattern(pattern: &MacroPattern, exprs: &[ExprKind], improper: bool) -> bool {
```
Comment (jrvidal) [inline crates/steel-core/src/parser/expander.rs:754]: :information_source: Special case for matching a rest pattern.

File: crates/steel-core/src/parser/expander.rs [hunk 5]
```diff
@@ -979,15 +996,25 @@ pub fn collect_bindings(
                         stop!(BadSyntax => "macro expected a list of values, not including keywords, found: {}", child)
                     }
                 } else {
-                    stop!(BadSyntax => "macro expected a list of values, not including keywords, found: {}", child)
+                    if let Some(pat) = non_list_match(&children) {
+                        collect_bindings(
+                            &[pat.clone()],
+                            &[child.clone()],
```
Comment (jrvidal) [inline crates/steel-core/src/parser/expander.rs:1002]: :warning: Some unfortunate cloning

**Review by mattwparas — APPROVED** [review]

**Pull request merged**